Lib/asyncio/tasks.py

```diff
@@ -564,62 +564,106 @@ async def _cancel_and_wait(fut):
         fut.remove_done_callback(cb)
 
 
-# This is *not* a @coroutine!  It is just an iterator (yielding Futures).
-def as_completed(fs, *, timeout=None):
-    """Return an iterator whose values are coroutines.
-
-    When waiting for the yielded coroutines you'll get the results (or
-    exceptions!) of the original Futures (or coroutines), in the order
-    in which and as soon as they complete.
+class _AsCompletedIterator:
+    """Doubles as an async iterator of as-completed tasks and futures
+    from a supplied set of awaitables and a plain iterator of
+    coroutines that resolve to results from the supplied awaitables as
+    their underlying tasks or futures complete.
+    """
+    def __init__(self, aws, timeout):
+        from .queues import Queue  # Import here to avoid circular import problem.
+        self._done = Queue()
+        self._timeout_handle = None
 
-    This differs from PEP 3148; the proper way to use this is:
+        loop = events.get_event_loop()
+        todo = {ensure_future(aw, loop=loop) for aw in set(aws)}
+        for f in todo:
+            f.add_done_callback(self._handle_completion)
+        if todo and timeout is not None:
+            self._timeout_handle = (
+                loop.call_later(timeout, self._handle_timeout)
+            )
+        self._todo = todo
+        self._todo_left = len(todo)
+
+    def __aiter__(self):
+        return self
+
+    def __iter__(self):
+        return self
+
+    async def __anext__(self):
+        if not self._todo_left:
+            raise StopAsyncIteration
+        assert self._todo_left > 0
+        self._todo_left -= 1
+        return await self._wait_for_one()
+
+    def __next__(self):
+        if not self._todo_left:
+            raise StopIteration
+        assert self._todo_left > 0
+        self._todo_left -= 1
+        return self._wait_for_one(resolve=True)
```

Comment:
No, it does not work. `.result()` should be called not when we do `next()`, but when we do `await` on its result.

It can be implemented as a separate helper:
```py
async def _resolve(coro):
    return (await coro).result()

def __next__(self):
    ...
    return _resolve(self._wait_for_one())
```
But a boolean parameter is simpler and perhaps adds less overhead. Of course, if this idiom would repeat many times, it would be more sense to introduce a helper.